src/System.Linq.Expressions/src/System/Linq/Expressions/Interpreter/ControlFlowInstructions.cs

```diff
@@ -478,6 +474,84 @@ public override string ToString()
         }
     }
 
+    internal sealed class EnterTryFaultInstruction : IndexedBranchInstruction
+    {
+        private TryFaultHandler _tryHandler;
+
+        internal void SetTryHandler(TryFaultHandler tryHandler)
+        {
+            Debug.Assert(tryHandler != null);
+            Debug.Assert(_tryHandler == null, "the tryHandler can be set only once");
+            _tryHandler = tryHandler;
+        }
+
+        public override int ProducedContinuations => 1;
+
+        internal EnterTryFaultInstruction(int targetIndex)
+            : base(targetIndex)
+        {
+        }
+
+        public override int Run(InterpretedFrame frame)
+        {
+            Debug.Assert(_tryHandler != null, "the tryHandler must be set already");
+
+            // Push fault. 
+            frame.PushContinuation(_labelIndex);
+
+            int prevInstrIndex = frame.InstructionIndex;
+            frame.InstructionIndex++;
+
+            // Start to run the try/fault blocks
+            var instructions = frame.Interpreter.Instructions.Instructions;
+
+            // C# 6 has no direct support for fault blocks, but they can be faked or coerced out of the compiler
+            // in several ways. Catch-and-rethrow can work in specific cases, but not generally as the double-pass
+            // will not work correctly with filters higher up the call stack. Iterators can be used to produce real
+            // fault blocks, but it depends on an implementation detail rather than a guarantee, and is rather
+            // indirect. This leaves using a finally block and not doing anything in it if the body ran to
+            // completion, which is the approach used here.
```

Comment:
(Strictly the jumping matter doesn’t really have an effect, since we block that separately and otherwise there'd be another way to get to the instructions bypassing this, but it's still nice to be as close to the semantics modelled as possible)